### ISSUE TITLE
DirServer.Watch does not signal receipt of the full tree

### ISSUE THREAD

**Issue by masiulaniec** [upspin/upspin#482]
The documentation for Watch states:

```
    // If order is WatchCurrent, the server first sends a sequence
    // of events describing the entire tree rooted at name. The Events are
    // sent in sequence such that a directory is sent before its contents.
    // After the full tree has been sent, the operation proceeds as normal.
```

How does one identify the event that completes the receipt of the full tree? I don't think it's possible right now. I need this for a code builder service I am working on.

**Comment by adg** [comment]
This is a good question, and I'm not sure of the answer.

The sequence number changes in #444 are not sufficient, I fear.

**Comment by robpike** [comment]
Maybe we should put a special code in the Event type. Receipt of an otherwise empty Event with that type would mean, "You have now received the state of the tree as it was when you issued the WatchCurrent operation."

**Comment by masiulaniec** [comment]
I was thinking along the same lines except I considered sending a duplicate Event for the root entry as an end-of-tree marker.

**Comment by adg** [comment]
>  duplicate Event for the root entry as an end-of-tree marker.

That doesn't sound too bad. Or maybe the root entry should just be the very last one sent? That's a lot easier to detect.

**Comment by masiulaniec** [comment]
Sending the root entry last would be incompatible with the current simple WatchCurrent API ("The Events are sent in sequence such that a directory is sent before its contents").

**Comment by adg** [comment]
@masiulaniec that's true, but if you get any events for Watch(path) then path must exist. We're talking about API changes here, anyway.

The rule could be, for `Watch(path, -1)` the entry for `path` will be the last event sent before the stream of tree updates.

**Comment by masiulaniec** [comment]
I now think that Rob's idea about sending empty Event is the best. In order to convince myself, I sketched a test program that uses Watch to create a mirror of an Upspin tree in a local directory. I tried both approaches and published them at:

  https://gist.github.com/masiulaniec/e0b3a133db1b808a5c80e92cda76c970

(For brevity, it lacks error checking and support for symlinks.)

The empty Event is strictly better: it's shorter, less tricky (no need for path.Clean, no need to decide between Mkdir and MkdirAll), faster (fewer mkdir syscalls), and needs no helper state (seenFull bool).

**Comment by adg** [comment]
Nice analysis. Sounds good to me.

**Comment by robpike** [comment]
I agree. Well done. Let's make it so.
